bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/TableOfContentsFilter.java

```diff
@@ -0,0 +1,295 @@
+        List<String> selectors = new ArrayList<>();
+        for(String className: classNames) {
+            for(int level = startLevel; level <= stopLevel; level++) {
+                selectors.add("." + className + " " + getHeadingTagName(level));
+                selectors.add(getHeadingTagName(level) + "." + className);
+            }
+        }
+        return StringUtils.join(selectors, ",");
+    }
+
+    /**
+     * Recursive method to create a nested list of TOC depending upon the heading levels of consecutive heading elements
+     * @param listTag - 'ul' for unordered list or 'ol' for ordered list
+     * @param headingElementsIterator - Iterator of list of heading elements to be included in TOC
+     * @param parentHeadingLevel - Heading level of the parent of the current nesting level
+     * @return Current nested TOC element containing all heading elements with levels >= parent heading level
+     */
+    private Element getNestedList(String listTag, ListIterator<Element> headingElementsIterator,
+                                  int parentHeadingLevel) {
+        if(!headingElementsIterator.hasNext()) {
+            return null;
+        }
+        Element list = new Element(listTag);
+        Element headingElement = headingElementsIterator.next();
+        Element listItem = getListItemElement(headingElement);
+        int previousHeadingLevel = getHeadingLevel(headingElement);
+        list.appendChild(listItem);
+        while(headingElementsIterator.hasNext()) {
+            headingElement = headingElementsIterator.next();
+            int currentHeadingLevel = getHeadingLevel(headingElement);
+            if(currentHeadingLevel == previousHeadingLevel ||
+                (currentHeadingLevel < previousHeadingLevel && currentHeadingLevel > parentHeadingLevel)) {
+                listItem = getListItemElement(headingElement);
+                list.appendChild(listItem);
+                previousHeadingLevel = currentHeadingLevel;
+            } else if(currentHeadingLevel > previousHeadingLevel) {
+                headingElementsIterator.previous();
+                list.children().last().appendChild(getNestedList(listTag, headingElementsIterator, previousHeadingLevel));
+            } else {
+                headingElementsIterator.previous();
+                return list;
+            }
+        }
+        return list;
+    }
+
+    /**
+     * Creates list item element('li') from the heading element.
+     * Adds an internal link on the 'li' element to the provided heading element.
+     * Adds 'id' attribute on heading element if not already present, using its text content.
+     * @param headingElement - DOM heading element
+     * @return Independent 'li' element, not attached to the DOM
+     */
+    private Element getListItemElement(Element headingElement) {
+        String id = headingElement.attr("id");
+        if("".contentEquals(id)) {
+            id = headingElement.text()
```

Comment:
Two heading elements with the same title will currently have the same generated ID. IDs should be unique.
To make the ID unique you could e.g. suffix the current ID with a counter that is incremented for each heading element.